cpmpy/solvers/gurobi.py

```diff
@@ -0,0 +1,376 @@
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: a CPMpy Model()
+        """
+        if not self.supported():
+            raise Exception(
+                "CPM_gurobi: Install the python package 'gurobipy' and make sure your licence is activated!")
+        import gurobipy as gp
+
+        # initialise the native solver object
+        self.env = gp.Env()
+        self.env.setParam("LogToConsole", 0)
+        self.env.setParam("OutputFlag", 0)
+        self.env.start()
+
+        self.gbi_model = gp.Model(env=self.env)
+        self._objective_value = None
+
+        # initialise everything else and post the constraints/objective
+        # it is sufficient to implement __add__() and minimize/maximize() below
+        super().__init__(name="gurobi", cpm_model=cpm_model)
+
+    def __add__(self, cpm_con):
+        """
+        Post a (list of) CPMpy constraints(=expressions) to the solver
+
+        Note that we don't store the constraints in a cpm_model,
+        we first transform the constraints into primitive constraints,
+        then post those primitive constraints directly to the native solver
+
+        :param cpm_con CPMpy constraint, or list thereof
+        :type cpm_con (list of) Expression(s)
+        """
+        # add new user vars to the set
+        self.user_vars.update(set(get_variables(cpm_con)))
+
+        # apply transformations, then post internally
+        # XXX chose the transformations your solver needs, see cpmpy/transformations/
+
+        cpm_cons = flatten_constraint(cpm_con)
+        cpm_cons = linearize_constraint(cpm_cons)
+
+        for con in cpm_cons:
+            self._post_constraint(con)
+
+        return self
+
+    def solve(self, time_limit=None, solution_callback=None, **kwargs):
+        """
+            Call the gurobi solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Arguments that correspond to solver parameters:
+            <Please document key solver arguments that the user might wish to change
+             for example: log_output=True, var_ordering=3, num_cores=8, ...>
+            <Add link to documentation of all solver parameters>
+        """
+        from gurobipy import GRB
+
+        if time_limit is not None:
+            raise NotImplementedError("TEMPLATE: TODO, implement time_limit")
+
+        # call the solver, with parameters
+        for param, val in kwargs.items():
+            self.gbi_model.setParam(param, val)
+
+        _ = self.gbi_model.optimize(callback=solution_callback)
+
+        my_status = self.gbi_model.Status
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = self.gbi_model.runtime
+
+        # translate exit status
+        if my_status == GRB.OPTIMAL and self._objective_value is None:
```

Comment:
self._objective_value is a value that you should fill in yourself. You should check with self.gbi_model somehow if it was an optimisation problem or not... perhaps check the gurobi sudoku model to see what they do there

Reply:
Gurobi models have a default objective value "Linexp(0)" so I could check if the objective equals that expression?